src/httpserver.h

```diff
@@ -27,168 +34,309 @@ static const int DEFAULT_HTTP_WORKQUEUE=64;
 
 static const int DEFAULT_HTTP_SERVER_TIMEOUT=30;
 
-struct evhttp_request;
-struct event_base;
-class CService;
-class HTTPRequest;
+enum HTTPRequestMethod {
+    UNKNOWN,
+    GET,
+    POST,
+    HEAD,
+    PUT
+};
 
-/** Initialize HTTP server.
- * Call this before RegisterHTTPHandler or EventBase().
- */
-bool InitHTTPServer(const util::SignalInterrupt& interrupt);
-/** Start HTTP server.
- * This is separate from InitHTTPServer to give users race-condition-free time
- * to register their handlers between InitHTTPServer and StartHTTPServer.
+/** Event handler closure.
  */
-void StartHTTPServer();
-/** Interrupt HTTP server threads */
-void InterruptHTTPServer();
-/** Stop HTTP server */
-void StopHTTPServer();
+class HTTPClosure
+{
+public:
+    virtual void operator()() = 0;
+    virtual ~HTTPClosure() = default;
+};
 
-/** Change logging level for libevent. */
-void UpdateHTTPServerLogging(bool enable);
+namespace http_bitcoin {
+using util::LineReader;
+using NodeId = SockMan::Id;
 
-/** Handler for requests to a certain HTTP path */
-typedef std::function<bool(HTTPRequest* req, const std::string &)> HTTPRequestHandler;
-/** Register handler for prefix.
- * If multiple handlers match a prefix, the first-registered one will
- * be invoked.
- */
-void RegisterHTTPHandler(const std::string &prefix, bool exactMatch, const HTTPRequestHandler &handler);
-/** Unregister handler for prefix */
-void UnregisterHTTPHandler(const std::string &prefix, bool exactMatch);
+// shortest valid request line, used by libevent in evhttp_parse_request_line()
+static const size_t MIN_REQUEST_LINE_LENGTH{strlen("GET / HTTP/1.0")};
+// maximum size of http request (request line + headers)
+// see https://github.com/bitcoin/bitcoin/issues/6425
+static const size_t MAX_HEADERS_SIZE{8192};
 
-/** Return evhttp event base. This can be used by submodules to
- * queue timers or custom events.
- */
-struct event_base* EventBase();
+class HTTPHeaders
+{
+public:
+    std::optional<std::string_view> Find(const std::string key) const;
+    void Write(const std::string key, const std::string value);
+    void Remove(const std::string key);
+    bool Read(util::LineReader& reader);
+    std::string Stringify() const;
+
+private:
+    std::map<std::string, std::string, util::CaseInsensitiveComparator> m_map;
```

Comment:
`std::map` would keep the entries sorted by the key in alphabetical order and lookup time is `O(log(number of elements in the map))` whereas `std::unordered_map` does not maintain an order and has a lookup time of `O(1)`. We do not need any order here, right? This probably is irrelevant since we are only going to put a bunch of entries in this map, but anyway, if there is no reason to use `std::map` I guess we should default to `std::unordered_map`.

(from commit e95c6f5b65 `http: Implement HTTPHeaders class`)